### PR TITLE
Repeatedly ask for farm name selection 

### PR DESCRIPTION
**Pull request by ranatrk**
Repeatedly ask for farm name selection if the farm doesnt have nodes to fit the required resources criteria in chatflows

### CODE REVIEW

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py [hunk 1]
```diff
@@ -145,29 +145,60 @@ def _nodes_distribute(self, number_of_nodes, farm_names):
...
+            if nodes_number > len(nodes):
+                farms_with_no_resources.append(farm_name)
+        return list(farms_with_no_resources)
+
+    def farm_names_get(self, number_of_nodes, bot, cru=None, sru=None, mru=None, hru=None, currency="TFT", message=""):
+        farms_message = f"Select 1 or more farms to distribute the {message} nodes on. If no selection is made, the farms will be chosen randomly"
+        empty_farms = set()
+        all_farms = j.clients.explorer.explorer.farms.list()
+        while True:
+            farms = j.sal.reservation_chatflow.farms_select(bot, farms_message)
```
Comment (grimpy) [inline JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py]: self

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py [hunk 2]
```diff
@@ -145,29 +145,60 @@ def _nodes_distribute(self, number_of_nodes, farm_names):
...
             nodes = list(nodes)
-            if free_to_use is True:
-                nodes = list(nodes)
-                nodes = filter(j.sal.zosv2.nodes_finder.filter_is_free_to_use, nodes)
-            elif free_to_use is False:
-                nodes = list(nodes)
-                nodes = filter(j.sal.zosv2.nodes_finder.filter_is_not_free_to_use, nodes)
-            return nodes
+            nodes = filter(j.sal.zosv2.nodes_finder.filter_is_not_free_to_use, nodes)
+        return list(nodes)
```
Comment (grimpy) [inline JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py:157]: why there is list here?

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py [hunk 3]
```diff
@@ -145,29 +145,60 @@ def _nodes_distribute(self, number_of_nodes, farm_names):
...
+            farms = j.sal.reservation_chatflow.farms_select(bot, farms_message)
+            farms_with_no_resources = j.sal.reservation_chatflow.farms_check(
+                1, farm_names=farms, cru=cru, sru=sru, mru=mru, hru=hru, currency=currency
+            )
+            if not farms_with_no_resources:
+                return farms
+            for farm_name in farms_with_no_resources:
+                empty_farms.add(farm_name)
+            if len(all_farms) == len(empty_farms):
+                raise StopChatFlow("No Farms available containing nodes that match the required resources")
```
Comment (grimpy) [inline JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py:190]: matches no?

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py [hunk 4]
```diff
@@ -145,29 +145,60 @@ def _nodes_distribute(self, number_of_nodes, farm_names):
...
+            farms_message = (
+                f"""The following farms dont have enough resources for {message}: """
```
Comment (grimpy) [inline JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py]: don't

**Review by grimpy — CHANGES_REQUESTED** [review]
nits so waleed can win

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py (hunk 3 above)
Comment (ranatrk) [reply JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py:190]: that is if all possible farms were checked and nothing matches, so they dont keep asking forever

File: JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py (hunk 2 above)
Comment (ranatrk) [reply JumpscaleLibs/sal/reservation_chatflow/reservation_chatflow.py:157]: @waleedhammam  got the following error as filter() doesnt return a list but a filter object
Screenshot from 2020-04-30 14-25-43

**Review by grimpy — APPROVED** [review]
LGTM

**Pull request merged**